---
title: "R Notebook"
output: html_notebook
editor_options: 
  chunk_output_type: console
---

```{r, echo=FALSE}
source("Rscripts/utils.R")
library(tidyverse)
library(data.table)
library(scales)
library(ggforce)
library(cowplot)
library(dplyr)
library(ggrepel)
library(glue)
library(patchwork)
library(ggpubr)
library(valr)
library(pals)


FONT_SIZE=6
my_grid = function(...){
    theme_minimal_grid(font_size=FONT_SIZE, ...)
} 

my_hgrid = function(...){
    theme_minimal_hgrid(font_size=FONT_SIZE, ...)
} 

my_vgrid = function(...){
    theme_minimal_vgrid(font_size=FONT_SIZE, ...)
} 

theme_no_x = function(...){
    theme(
        axis.title.x=element_blank(),
        axis.text.x=element_blank(),
        axis.ticks.x=element_blank()
    )
}

my_read_bed = function(...){
    df=fread(...)
    colnames(df)[1:3]=c("chrom", "start", "end")
    df
}

in_file="results/GM12878_FDR/FDR-peaks/FDR-FIRE-peaks.bed.gz"
out_file="Figures/peaks-vs-percentage.pdf"
encode=my_read_bed("data/ENCODE3_consensus_DHS_ENCFF503GCK.tsv")
sds=my_read_bed("data/T2T_SDs.bed")
sds=my_read_bed("config/annotations/SDs.merged.hg38.bed.gz")
tss=my_read_bed("data/gencode.v42.annotation_TSS.gff3")
```

```{r}
#  compare_fdr_and_dnase/ucsc/ENCFF960FMM.bedGraph
dnase=my_read_bed("data/bedgraph_annotations/ENCFF960FMM_dnase_signal.bed")
colnames(dnase)[4] = "dnase_sig"
dnase

```


```{r}
df=fread(in_file)
df$peak_cov = df$coverage
df$acc_percent = df$fire_coverage/df$peak_cov
colnames(df)[1:5]=c("chrom", "start","end","ostart","oend")


df = df %>%
    arrange(-acc_percent) %>%
    mutate(
        n=seq(n()),
        min_percent_acc=min(acc_percent)
    ) %>%
    arrange(-n) %>%
    mutate(
        Mbp=cumsum(end-start)/1e6,
        group = floor(acc_percent * 20) / 20,
        group = case_when(
            group == 1 ~ 0.95,
            TRUE ~ group,
        )
    ) %>%
    bed_map(encode,
        encode_count=length(core_end),
        encode_anno = paste0(unique(component), collapse=";")
    ) %>%
    bed_map(sds, sd_count=length(end)) %>%
    bed_map(dnase, dnase_max=max(dnase_sig)) %>%
    bed_map(tss, TSS=length(V4)) %>%
    replace_na(list(encode_count = 0, sd_count=0, TSS=0)) %>%
    arrange(-acc_percent, -n) %>%
    data.table()
head(df)
tail(df)
dim(df)
df %>%
    group_by(encode_anno) %>% summarise(n())
table(df$group)
```

```{r}
percent_encode = df %>% 
    group_by(group) %>%
    summarise(
        `% in\nENCODE` = 100*mean(encode_count>0),
        `% in\nSDs`=100*mean(sd_count>0),
        `% in\nTSS`=100*mean(TSS>0),
        #`% of genome\n(c)`=100*sum(end-start)/3.1e9,
        `% of\ngenome\n(cumulative)`=100*max(Mbp*1e6)/3.1e9,
        `# of\npeaks`=n(),
        acc_percent=unique(group)+0.025,
        Mbp = sum(end-start)/1e6,
    ) %>%
    filter(`# of\npeaks` > 2)

cur_colors = c(
            `% in\nENCODE`="darkgreen",
            `% in\nTSS`="darkblue",
            `% in\nSDs`="darkred",
            `% of\ngenome\n(cumulative)`="darkgray",
            `# of\npeaks`="darkcyan"
            )
cur_colors
percent_encode_long = percent_encode %>%
    pivot_longer(
        cols=names(cur_colors)
        )

percent_encode_long

percent_plots = percent_encode_long %>%
    mutate(
        name=factor(name, levels=names(cur_colors))
    ) %>%
    ggplot(aes(x=acc_percent, y=value, group=name, fill=name)) +
    geom_bar(stat="identity", color="black")+
    geom_text(
        aes(
            x=acc_percent,
            label=sub("\\.0+$", "", comma(value, accuracy=0.01)),
        ),
        #min.segment.length=1,
        #direction="y",
        #alpha=0.5,
        vjust=-0.3,
        #color="white",
        #nudge_y=0.01,
        size=1
    ) +
    scale_y_continuous("", label=comma) +
    scale_x_continuous(
        "",
        breaks=seq(5,100,5)/100,
        label=rep("",20),
    )+
    scale_fill_manual(values=cur_colors) + 
    facet_col(~name, scales="free_y", strip.position = "left") + 
    my_grid() +
    coord_cartesian(xlim=c(0,1), clip=FALSE)+
    theme_no_x() +
    theme(legend.position="None")

my_ggsave("Figures/peaks-vs-percent.pdf", height=3, width=5)
```

```{r}
x = df %>% 
    separate_rows(encode_anno, sep=";") %>%
    group_by(group, encode_anno) %>%
    summarise(
        count=n()   
    ) %>%
    group_by(
        group
    ) %>%
    filter(sum(count)>20) %>%
    mutate(
        percent=100*count/sum(count)
    )
encode %>%
    group_by(component) %>%
    summarise(count=n(),percent = sum(end-start)/1e6) %>%
    mutate(
        #group="control",
        encode_anno=component,
        percent = percent/sum(percent)*100
    ) %>%
    arrange(percent) %>%
    merge(x, by="encode_anno") %>%
    filter(count.y>300) %>%
    data.table() %>%
    mutate(
        lfc = log2(percent.y) - log2(percent.x)
    ) %>%
    arrange(-lfc)


plot_encode_anno=x %>% 
    filter(!is.na(encode_anno)) %>%
    ggplot(
        aes(
            x=group+0.025,
            weight=count,
            fill=encode_anno,
        )
    ) +
    geom_bar(
        position="fill",
        width=0.0495,
    ) +
    scale_x_continuous(
        "",
        breaks=seq(5,100,5)/100,
        label=rep("",20),
    )+
    scale_y_continuous("")+
    scale_fill_manual(
        "", 
        values=as.vector(glasbey(length(unique(encode$component)))),
        guide=guide_legend(nrow=2)
    )+
    coord_cartesian(xlim=c(0,1))+
    my_grid()+
    theme(
        legend.text=element_text(size=3),
        legend.key.size=unit(1, "mm"),
        legend.position="top",
        plot.margin = unit(c(0, 0, 0, 0), "cm"),
    )
kk=plot_encode_anno+plot_spacer()+percent_plots+plot_layout(heights = c(1,-0.3, 3))
my_ggsave("Figures/peaks-vs-encode-annotations.pdf", height=4, width=6, plot=kk)
```


```{r}
by_5_per = df %>%
    group_by(group) %>%
    slice_max(order_by = n, n = 1)

lookup=df$Mbp
names(lookup)=df$n
#lookup
my_scale_function=function(x){
    lookup[x]
}

pecdf=df %>%
    arrange(-acc_percent, -n) %>%
    #filter((n+1)%%100==0) %>%
    ggplot(aes(x=acc_percent, y=n)) +
    geom_line()+
    geom_text_repel(
        data = df %>% filter(n==max(n)),
        aes(
            x=min_percent_acc,
            label=paste(
                "Limit of detection", percent(min_percent_acc, accuracy=0.01),
                "\n# peaks", comma(n)
            ),
        ),
        min.segment.length=0,
        segment.size=0.2,
        direction="y",
        size=1.5,
        nudge_y=-1.0,
        #nudge_x=0.1,
    ) +
    geom_text_repel(
        data=by_5_per,
        aes(
            x=acc_percent,
            label=paste(
                comma(n)
            ),
        ),
        min.segment.length=0,
        direction="x",
        nudge_x=0.1,
        segment.size=0.1,
        size=1
    ) +
    scale_y_continuous(
        "# of regulatory elements\nin the genome",
        trans="log10",
        label=comma,
         # Add a second axis and specify its features
       #sec.axis = sec_axis(
       #     ~my_scale_function(.),
       #     name="Mbp",
       #     label=comma,
        #)
    ) + 
    annotation_logticks(side="l")+
    scale_x_continuous(
        "Minimum % of fibers that are accessible",
        breaks=seq(5,100,5)/100,
        label=percent,
        #guide = guide_axis(n.dodge = 2),
    ) +
    my_grid() +
    coord_cartesian(xlim=c(0,1), ylim=c(100,NA)) 



p=plot_encode_anno+plot_spacer()+percent_plots+pecdf+plot_layout(heights = c(1,-0.5, 3, 2))
my_ggsave("Figures/peaks-vs-percent.pdf", height=4, width=5)
```


#dnase
```{r}
df %>%
    filter(dnase_max > 0) %>%
    ggplot(aes(x=score, y=dnase_max+1)) +
    geom_hex(bins=30) +
    scale_fill_distiller("", palette = "Spectral", trans="log10") +
    stat_cor()+
    geom_smooth(method="lm", se=F) +
    #facet_wrap(~autosome, ncol=2)+
    #facet_wrap(~group)+
    scale_x_continuous(
        "FIRE peak score",
        labels=comma,
        #trans="log10",
    ) +
    scale_y_continuous(
        "DNase signal",
        labels=comma,
        trans="log10",
    ) + 
    my_grid()
my_ggsave("Figures/DNase-vs-FIRE.pdf", height=3, width=4)
```



# number of regulatory elements per cell
```{r}
comma(sum(df$acc_percent)*2)
```


```{r}
bp = fread("data/bedgraph_annotations/bo.10bp.bed.gz", nThread=8)

bp[, c("chrom", "start", "end") := tstrsplit(key, "_", fixed=TRUE, type.convert=TRUE)]
bp$mid = floor((bp$start + bp$end)/2)
bp$position = bp$st - bp$mid
dim(bp)/1e6
bp

df_bp = merge(bp, df[, c("chrom", "start", "end", "group")], by=c("chrom", "start", "end")) %>%
    replace_na(list(ENCFF658WKQ_DNase = 0)) %>%
    pivot_longer(
        cols = starts_with("EN", ignore.case = FALSE),
        names_to= "exp",
        values_to="coverage",
    ) %>%
    filter(!is.na(coverage)) %>%
    data.table
df_bp
```
```{r}
width=500
width=1000
smooth=10
#df_bp %>% group_by(group) %>% summarise(length(unique((key))))
p.df = df_bp %>% 
    filter(chrom != "chrY", chrom != "chrX") %>%
    filter(position > -width & position < width) %>%
    filter(exp=="ENCFF658WKQ_DNase") %>%
    mutate(
        position = smooth * floor(position/smooth),
    ) %>%
    group_by(position, exp, group) %>%
    summarise(
        n_sites = length(unique(key)),
        value = sum(coverage),
    ) %>%
    group_by(group, exp) %>%
    mutate(
        n_sites=mean(n_sites),
        value2 = value/mean(n_sites),
    ) %>%
    filter(n_sites>100) %>%
    arrange(exp, group, position) 

mmin=min(p.df$group)
mmax=max(p.df$group)
labels = seq(mmin,mmax,0.05)
labels
z = p.df %>% 
    ggplot(
        aes(
            x=position, 
            y=value2,
            #weight=coverage,
            #color=factor(`group`),
            color=`group`-0.5,
            fill=NULL, 
            group=group
        )
    ) +
    geom_line(linewidth=0.2)+
    #geom_density(alpha=0.8, linewidth=.3) +
    #facet_wrap(~exp, scales="free")+
    scale_x_continuous("Relative position") +
    scale_y_continuous("DNase read count", trans="log10") + annotation_logticks(side="l")+
    scale_colour_steps2("", breaks=labels-0.5,  labels=labels) +
    my_grid() +
    theme(
        #legend.position="top",
        legend.key.size=unit(0.2, "inch")
    )
my_ggsave("Figures/DNase-over-FIRE-peaks.pdf", height=2, width=2.75)
```

```{r}
smooth = 10
z=df_bp %>%
    filter(group == 0.20) %>%
    #head(1e6) %>%
    #filter(group > 0.50) %>%
    filter(!is.na(coverage)) %>%
    filter(position > -width & position < width) %>%
    #group_by(group, position, exp) %>%
    #summarise(
    #    n_sites = length(unique(key)),
    #    coverage = sum(coverage),
    #) %>%
    ggplot(
        aes(
            x=position, 
            y=coverage,
            #weight=coverage,
            #color=factor(`group`),
            #color=`group`-0.5,
        )
    ) +
    #geom_line(alpha=0.1)+
    geom_hex(bins=300) + scale_fill_distiller("", palette = "Spectral", trans="log10")+
    #geom_density(alpha=0.8, linewidth=.3) +
    facet_wrap(~exp, scales="free")+
    scale_x_continuous("Relative position") +
    #scale_y_continuous("Relative position", trans="log10") +
    scale_y_continuous("Relative position", limits=c(0,50)) +
    #scale_colour_steps2(n.breaks=20, nice.breaks=TRUE) +
    my_grid() 
my_ggsave("Figures/Other-DNase-over-FIRE-peaks.pdf", height=3, width=5)
```

```{r}
p.df[p.df$position==-width& p.df$exp=="ENCFF658WKQ_DNase",]
p.df[p.df$position==width-smooth& p.df$exp=="ENCFF658WKQ_DNase",]

df_bp %>%
    filter(group == 0.05) 
```



# Figure out 5mC over low peaks
```{r}
# this is slow
low_acc = df[acc_percent < 0.25 & FDR <= 0.01]
low_acc %>%
    sample_n(1000) %>%
    select(chrom,start,end) %>%
    fwrite("temp/low_acc_peaks.bed", row.names=F, col.names=F, quote=F, sep="\t")
low_acc
system("ft center -t $(nproc) -d 500 -r ../k-mer-variant-phasing/results/GM12878/GM12878.haplotagged.bam temp/low_acc_peaks.bed | rg '5mC|type'| bgzip -@ $(nproc) > temp/low_acc_center.tbl.gz")
```

```{r}
center=fread("temp/low_acc_center.tbl.gz") %>%
    select(-subset_sequence) %>%
    filter(centered_position_type=="5mC") %>%
    filter(centered_qual >= 256/2) %>%
    data.table()
center

system("tabix -R temp/low_acc_peaks.bed results/GM12878_FDR/fiber-calls/FIRE.bed.gz > temp/low_acc_fires.bed")
fires = my_read_bed("temp/low_acc_fires.bed")

center$fire = FALSE
center[query_name %in% fires$V4, "fire"] = TRUE
sum(center$fire)

```

```{r}
sum_data = center %>% 
    filter(centered_start < 250 & centered_start > 0) %>%
    group_by(chrom, centering_position, fire) %>%
    summarise(
        count = n()/length(unique(query_name)),
        w_count = sum(centered_qual)/length(unique(query_name)) * 1/256
    ) %>%
    data.table()

sum_data %>%
    ggplot(
        aes(color=fire, x=count)
    ) +
    facet_col(~fire, scales="free_y") + 
    geom_histogram(
        binwidth=0.25,
    ) +
    scale_x_continuous(
        "# of 5mC events in the peak",
        #trans="log10",
        label=comma
    )+
    theme_minimal_grid()


center %>% 
    #filter(centered_start < 250 & centered_start > 0) %>%
    group_by(chrom, centering_position, fire) %>%
    summarise(
        average_5mC_count = n()/length(unique(query_name))
    ) %>% 
    pivot_wider(
        id_cols=c("chrom", "centering_position"),
        names_from=fire, values_from=average_5mC_count
    ) %>%
    ggplot(aes(x=`FALSE`, y=`TRUE`)) +
    stat_cor()+
    geom_hex(bins=30) +  scale_fill_distiller("", palette = "Spectral", trans="log10") +
    geom_smooth(se=F, method="lm")+
    scale_x_continuous("Average number of 5mC events in reads without FIRE MSPs", trans="log10")+
    scale_y_continuous("Average number of 5mC events in reads with FIRE MSPs", trans="log10")+
    annotation_logticks(side="lb")+
    theme_minimal_grid()


my_ggsave("Figures/5mC_in_FIRE_fibers.pdf")
sum_data %>% 
    group_by(fire) %>%
    summarise(
        mean(count),
        median(count),
        mean(w_count)
    )
```